src/go/vtunnel/pkg/vmsock/conn_windows.go

```diff
@@ -0,0 +1,181 @@
+	if err != nil {
+		return nil, fmt.Errorf("vsockListen, could not parse Hyper-v service GUID: %v", err)
+	}
+
+	addr := hvsock.Addr{
+		VMID:      vmGuid,
+		ServiceID: svcPort,
+	}
+
+	return hvsock.Listen(addr)
+}
+
+// vmGuid retrieves the GUID for a correct hyper-v VM (most likely WSL).
+// It performs a handshake with a running peer process in the WSL distro
+// to make sure we establish the AF_VSOCK connection with a right VM.
+func vmGuid() (hvsock.GUID, error) {
+	key, err := registry.OpenKey(
+		registry.LOCAL_MACHINE,
+		`SOFTWARE\Microsoft\Windows NT\CurrentVersion\HostComputeService\VolatileStore\ComputeSystem`,
+		registry.ENUMERATE_SUB_KEYS)
+	if err != nil {
+		return hvsock.GUIDZero, fmt.Errorf("could not open registry key, is WSL VM running? %v", err)
+	}
+	defer key.Close()
+
+	names, err := key.ReadSubKeyNames(0)
+	if err != nil {
+		return hvsock.GUIDZero, fmt.Errorf("machine IDs can not be read in registry: %v", err)
+	}
+	if len(names) == 0 {
+		return hvsock.GUIDZero, errors.New("no running WSL VM found")
+	}
+
+	found := make(chan hvsock.GUID, len(names))
+	done := make(chan bool, len(names))
+	defer close(done)
+
+	for _, name := range names {
+		vmGuid, err := hvsock.GUIDFromString(name)
+		if err != nil {
+			logrus.Errorf("invalid VM name: [%s], err: %v", name, err)
+			continue
+		}
+		go handshake(vmGuid, found, done)
+	}
+	return tryFindGuid(found)
+}
+
+// tryFindGuid waits on a found chanel to receive a GUID until
+// deadline of 10s is reached
+func tryFindGuid(found chan hvsock.GUID) (hvsock.GUID, error) {
+	bailOut := time.After(time.Second * timeoutSeconds)
+	for {
+		select {
+		case vmGuid := <-found:
+			return vmGuid, nil
+		case <-bailOut:
+			return hvsock.GUIDZero, errors.New("could not find vsock-peer process on any hyper-v VM(s)")
+		}
+	}
+}
+
+// handshake attempts to perform a handshake by verifying the seed with a running
+// af_vsock peer in WSL distro, it attempts once per second
+func handshake(vmGuid hvsock.GUID, found chan<- hvsock.GUID, done <-chan bool) {
+	svcPort, err := hvsock.GUIDFromString(winio.VsockServiceID(PeerHandshakePort).String())
+	if err != nil {
+		logrus.Errorf("hostHandshake parsing svc port: %v", err)
+	}
+	addr := hvsock.Addr{
+		VMID:      vmGuid,
+		ServiceID: svcPort,
+	}
+
+	attempInterval := time.NewTicker(time.Second * 1)
+	attempt := 1
+	for {
+		select {
+		case <-done:
+			logrus.Infof("attempt to handshake with [%s], goroutine is terminated", vmGuid.String())
+			return
+		case <-attempInterval.C:
+			conn, err := hvsock.Dial(addr)
+			if err != nil {
+				attempt++
+				logrus.Debugf("handshake attempt[%v] to dial into VM, looking for vsock-peer", attempt)
+				continue
+			}
+			seed, err := readSeed(conn)
+			if err != nil {
+				logrus.Errorf("hosthandshake attempt to read the seed: %v", err)
+			}
+			if err := conn.Close(); err != nil {
+				logrus.Errorf("hosthandshake closing connection: %v", err)
+			}
+			if seed == SeedPhrase {
```

Comment:
Correct, the current behavior is, that we call `handshake` against all the hyper-v VMs that is found in the registry since we don't know which one is the WSL. The handshake is triggered in parallel with some intervals and timeouts to accommodate for the peer process startup time/delay. If a peer process is found we get a seed phrase if not the routine will exist under the following conditions:
1) seed not found and timeout reached
2) peer is found and seed is matched in one goroutine which can close all other goroutines.

Reply:
Right, but the current code just loops again if the seed is incorrect.

Reply:
why would the seed ever be incorrect?? it's a shared string constant between two processes. Both processes are literally sharing the same string constant.

Reply:
If it's a _different_ VM with some other thing listening on that vsock port (not ours)? Isn't that the whole point of having a seed?